Applications/Spire/Source/TimeAndSales/TimeAndSalesWindow.cpp

```diff
@@ -1,40 +1,191 @@
+      }
+    }
+  }
+  return Window::eventFilter(watched, event);
+}
+
+int TimeAndSalesWindow::get_row_height() const {
+  if(m_table_model->get_row_size() == 0) {
+    return get_height(m_factory->make()->get_current()->get().get_font());
+  }
+  return m_table_view->get_body().get_item({0, 0})->height();
+}
+
+void TimeAndSalesWindow::make_table_header_menu() {
+  if(m_table_header_menu) {
+    return;
+  }
+  m_table_header_menu = new ContextMenu(*this);
+  auto add_sub_menu = [&] (int column, const QString& name, bool checked) {
+    auto model = m_table_header_menu->add_check_box(name);
+    model->set(checked);
+    model->connect_update_signal(std::bind_front(
+      &TimeAndSalesWindow::on_header_item_check, this, column));
+  };
+  auto header_items = m_table_view->get_header().get_items();
+  for(auto i = 0; i < header_items->get_size() - 1; ++i) {
+    add_sub_menu(i, header_items->get(i).m_name,
+      m_table_view->get_header().get_item(i)->isVisible());
+  }
 }
 
 void TimeAndSalesWindow::on_current(const Security& security) {
   setWindowTitle(to_text(security) + " " + QString(0x2013) + " " + TITLE_NAME);
-  m_transition_view->set_status(TransitionView::Status::LOADING);
+  m_transition_view->set_status(TransitionView::Status::NONE);
+  m_table_model->set_model(m_model_builder(security));
+  auto& header = m_table_view->get_header();
+  m_table_model->load_history(
+    (m_body->height() - header.height()) / get_row_height());
+  auto& properties = m_factory->make()->get_current()->get();
+  update_style(*m_table_view, [&] (auto& style) {
+    auto header_item_selector =
+      Any() > is_a<TableHeader>() > is_a<TableHeaderItem>();
+    auto body_item_selector =
+      Any() > is_a<TableBody>() > Row() > is_a<TableItem>();
+    style.get(header_item_selector > TableHeaderItem::Label()).
+      set(Font(properties.get_font()));
+    style.get(body_item_selector > (is_a<TextBox>() || is_a<DecimalBox>())).
+      set(Font(properties.get_font()));
+    for(auto i = 0; i < BBO_INDICATOR_COUNT; ++i) {
+      auto indicator = static_cast<BboIndicator>(i);
+      apply_indicator_style(style, get_bbo_indicator_selector(indicator),
+        properties.get_highlight_color(indicator));
+    }
+  });
+}
+
+void TimeAndSalesWindow::on_header_item_check(int column, bool checked) {
+  m_table_view->get_header().get_item(column)->setVisible(checked);
+}
+
+void TimeAndSalesWindow::on_table_operation(
+    const TableModel::Operation& operation) {
+  visit(operation,
+    [&] (const TableModel::AddOperation& operation) {
+      auto row = m_table_view->get_body().get_item(
+        {operation.m_index, 0})->parentWidget();
+      row->setDisabled(true);
+      auto indicator = m_table_model->get_bbo_indicator(operation.m_index);
+      if(indicator == BboIndicator::UNKNOWN) {
+        match(*row, UnknownIndicator());
+      } else if(indicator == BboIndicator::ABOVE_ASK) {
+        match(*row, AboveAskIndicator());
```

Comment:
This approach is a good start but it's too fragile and depends on implementation details.

Try an approach where you apply the indicator styling on the cell, and then use the parent selector to select the ```Row()```. Something like:

```AtAskIndicator() < is_a<TableItem>() < Row()```